src/js/trove/matrix.js

```diff
@@ -0,0 +1,1310 @@
+            runtime.checkArgsInternal1("Matrix", "transpose", self, annMatrix);
+            new_arr = new Array(self.$l);
+            for (var i = 0; i < self.$h; i++) {
+                for (var j = 0; j < self.$w; j++) {
+                    new_arr[get1dpos(j, i, self.$h)] = get1dElem(self, i, j);
+                }
+            }
+            return createMatrixFromArray(self.$w, self.$h, new_arr);
+
+        }
+        //stacks matrices vertically
+        var stackMatrix = runtime.makeFunction(function (self, other) {
+            runtime.ffi.checkArity(2, arguments, "stack-mat", false);
+            runtime.checkArgsInternal2("Matrix", "stack-mat", self, annMatrix, other, annMatrix);
+            if (self.$w != other.$w) {
+                runtime.ffi.throwMessageException("Matrices need to have same width to be stacked ")
+            } else {
+                new_arr = new Array(self.$l + other.$l);
+                duplicateArray(self, 0, self.$l, new_arr, 0);
+                duplicateArray(other, 0, other.$l, new_arr, self.$l);
+                return createMatrixFromArray(self.$h + other.$h, self.$w, new_arr);
+            }
+        }, "stack-mat");
+
+        //multiplies each element in matrix by num 
+        var scaleMatrix = runtime.makeFunction(function (self, num) {
+            runtime.ffi.checkArity(2, arguments, "scale", false);
+            runtime.checkArgsInternal2("Matrix", "scale", self, annMatrix, num, runtime.Number);
+            new_mtrx = duplicateMatrix(self);
+            for (var i = 0; i < new_mtrx.$l; i++) {
+                new_mtrx.$underlyingMat[i] = runtime.times(new_mtrx.$underlyingMat[i], num);
+            }
+            return new_mtrx;
+        }, "scale");
+
+        //Applies f to each element in the matrix
+        var mapMatrix = function (self, f) {
+            arity(2, arguments, "map-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "map-mat", self, annMatrix, f, runtime.Function);
+            new_mtrx = duplicateMatrix(self);
+    
+            function helper(i,j) {
+                return runtime.safeCall(
+                     function() { return f.app(runtime.makeNumber(i),runtime.makeNumber(j),get1dElem(self,i,j)) ;} ,
+                     function(result) { 
+                        new_mtrx.$underlyingMat[get1dpos(i, j, self.$w)] = result ;
+                        if((i+1)*(j+1) ==  self.$l) { 
+                            return new_mtrx ; 
+                        } else {  
+
+                        nj = (j+1 ) % self.$w ;
+                        ni = i ; 
+                        if (nj == 0 ) { ni++ ; }
+                        } ; 
+                        return helper(ni,nj) ; 
+                     },"map-mat"
+                ) ; 
+            }
+            return helper(0,0); 
+        }
+        //Returns a matrix given by (n1,n2) in top left corner and (n3,n4) in bottom right 
+        var getSubmatrix = function (self, n1, n2, n3, n4) {
```

Comment:
I've handled the submatrix part. 

In our initial discussions, we were recommended to use a 1-d array representation. The idea was that with a 1-d array it would be easier to extend to Tensors (n-dimensional matrices ex : a matrix of dimensions(3,4,3)). 

I do agree there are a lot of brittle calculations involved right now. Do you think it would be best to switch to a 2d representation ?

Reply:
I'm not sure whether the current design generalizes any better or worse than having a multi-dimensional array would.  The one thing that does work "trivially" right now is the `reshape` operation, since you don't actually do anything to the array of values.  But I don't know how reshape would generalize to n dimensions anyway.

My main concern about a bunch of these things is I'd prefer robustness over sheer speed: we need to be able to understand and maintain this code after you finish with it! :)

Reply:
I agree that the overhead of having a 1d representation,  overweighs the small benefits (`reshape,transpose`).   I can also see future maintainers having problems with it. 

So do we want to change it ? If required it can be done without much work